src/libs/sfdk/virtualmachine.cpp

```diff
@@ -402,6 +402,67 @@ void VirtualMachine::setStorageSizeMb(int storageSizeMb, const QObject *context,
     });
 }
 
+void Sfdk::VirtualMachine::currentFreeSize(const QObject* context, const Functor<int> &functor)
+{
+    auto vmStarted = [=](bool ok) {
+        if (ok) {
+            auto *runner = BatchComposer::enqueue<RemoteProcessRunner>("get-current-free-size",
+                QString::fromLatin1("df --output=avail / | tail -n 1"), sshParameters());
+            connect(runner, &RemoteProcessRunner::done, context, [=](bool ok) {
+                int freeSize = -1;
+                if (ok) {
+                    QString output = QString::fromUtf8(runner->sshRunner()->readAllStandardOutput());
+                    bool ok;
+                    freeSize = output.toInt(&ok);
+                    if (!ok)
+                        freeSize = -2;
+                }
+                if (context)
+                    functor(freeSize);
+            });
+        }
+    };
+
+    if (isOff()) {
+        connectTo(NoConnectOption, this, vmStarted);
```

Comment:
This is problematic. So far the API is explicit about starting/stopping VMs, so we should not break that. We should either cache the free space, so that the VM does not need to be started in order to query it (preferred) or assert that the VM is started (connected) when this function is used. With the latter option we would then need to shut the VM down for the resize and that would hit similar issue. So the former option seems to be not only the preferred but also the only viable option.

Reply:
I find caching problematic as well, as pretty much everything we do in the vm affects the free space. We should update the cache every time we shut down the vm, and I don't like the idea of doing anything extra when shutting down.

Reply:
That was my first idea too, to simply save it at shutdown. In addition to that, most actions go through mb2 and sdk-manage, so those two could trigger that too easily. I wouldn't be too scare of getting out of sync as much as to actually break things easily.

Reply:
I'll try with the "save at shutdown" approach. It should be relatively easy to implement.